enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,305 @@
+This information could be correlated to e2e failures, alerts, events, logs, or other information feeds.
+
+## Motivation
+
+This attempts to identify failures of particular connections in a black-box manner that is independent of SDN, cloud,
+or cluster ingress.
+
+### Goals
+
+1. be independent of SDN, cloud, or cluster ingress
+2. Identify connection failure between each kas and each oas endpoint
+3. Identify connection failure between each kas and each etcd
+4. Identify connection failure between each kas and oas service IP
+5. Identify connection failure between each oas and each kas host IP
+6. Identify connection failure between each oas and each kas endpoint IP
+7. Identify connection failure between each oas and each etcd
+
+### Non-Goals
+
+1. DNS resolution, that would be distinct.
+
+## Proposal
+
+1. One instance of a new CR per point to point connection.
+2. Detailed status, probably some kind of latency information.
+3. It should be possible to clean up garbage after a couple hours of inactivity.
+4. Don't delete aggressively, on upgrades, we have different endpoints
+5. Create a binary that can take multiple destinations paired with instances to write to.
+6. Use that binary in containers included in each kas and oas (or other interesting) pod.
+
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+#### RelatedEndpoint Custom Resource
+
+```go
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConditionCheck references a "interesting" endpoint for a given pod.
+type ConditionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   ConditionCheckSpec   `json:"spec"`
+	Status ConditionCheckStatus `json:"status"`
+}
+
+type ConditionCheckSpec struct {
+	// Selector for pods which should check this condition
+	Selector *metav1.LabelSelector      `json:"selector"`
+	Endpoint ConditionCheckSpecEndpoint `json:"endpoint"`
+}
+
+type ConditionCheckSpecEndpoint struct {
+	// URL is the url of the end point
+	URL string `json:"url"`
+	// CACertFile, if specified, will be used to validate server certs
+	CACertFile string `json:"url,omitempty"`
+	// CertFile, if specified, will be sent to the endpoint as a client cert
+	CertFile string `json:"url,omitempty"`
+	// KeyFile, if specified, will be used when CertFile is also specified
+	KeyFile string `json:"url,omitempty"`
```

Comment:
this feels like client auth? should we expect that this network-check run on no-auth endpoints like the health check.

because why would we only support only one type of authentication? If we want to support client auth, i think we should make it configurable structure not embedded like this which is difficult to exand later on.

Reply:
I've changed to limit scope to just the tcp connection.